tiledb/sm/query/readers/aggregators/iaggregator.h

```diff
@@ -0,0 +1,104 @@
+/**
+ * @file   iaggregator.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class IAggregator.
+ */
+
+#ifndef TILEDB_IAGGREGATOR_H
+#define TILEDB_IAGGREGATOR_H
+
+#include "tiledb/common/status.h"
+#include "tiledb/sm/enums/layout.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class ArraySchema;
+class QueryBuffer;
+class AggregateBuffer;
+class IAggregator;
+
+typedef std::unordered_map<std::string, shared_ptr<IAggregator>>
+    DefaultChannelAggregates;
+
+class IAggregator {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  virtual ~IAggregator() = default;
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /** Returns the field name for the aggregator. */
+  virtual std::string field_name() = 0;
+
+  /** Returns if the aggregate needs to be recomputed on overflow. */
+  virtual bool need_recompute_on_overflow() = 0;
```

Comment:
Overflow of what?

Reply:
This is a var size overflow. When we run any query, the user gives us a var buffer when they request a variable length field. When we copy the data to the var buffer, we might realize that we can fit less cells than we previously thought. We then need to scale the results back for any fields that were previously copied as all buffers need to return the same number of cells. We made the decision to use, for now, the same structure to save progress for both the copy operation and the aggregate operation. For a sum operation, that means that if we already summed some data then we later run into a var sized overflow, we need to reprocess the sum so that we don't include more cells in the sum than what are included in the regular buffers. That way cells are not counted twice when we continue the INCOMPLETE query. For some aggregates (MIN/MAX for example), this doesn't matter as the aggregation is an absolute and processing cells twice will yield the same result.